### ISSUE TITLE
Potential False Detection of Issue ST06 - structure.column_order

### ISSUE THREAD

**Issue by Asma280** [sqlfluff/sqlfluff#4859]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

In my code , SQLfluff raised the following issue 

> L:   2 | P:   5 | ST06 | Select wildcards then simple targets before calculations
>                        | and aggregates. [structure.column_order]
> 

Here is the code : 

```sql
    select
        case_id
        , max(case when new_value_string = 'AA' then event_history_at end) as kick_off_at
        , max(case when new_value_string = 'BB' then event_history_at end) as moved_to_cs_at
        , max(case when new_value_string = 'CC' then event_history_at end) as account_delivered_at
        , min(case when modified_field = 'DD' then event_history_at end) as assigned_at
        , min(case when new_value_string = 'EE' then event_history_at end)::date as lost_date
    from {{ ref('stg_salesforce__casehistory') }}
    group by 1
``` 

### Expected Behaviour

Sqlfluff raise an error even tho the aggregates columns are placed at the end of the code.
If it is to be kept in this way, it would be better to indicate the correct line number, as selecting line 2 can be confusing. This enhancement would provide clearer error messages and improve the overall user experience when encountering such scenarios.

### Observed Behaviour

After the observation, the issue is due to `::date `

Now on sqlfluf
❌ `, min(case when new_value_string = 'Lost' then event_history_at end)::date as lost_date`
✅ `, min(case when new_value_string = 'Lost' then event_history_at::date end) as lost_date`

the first line (❌) should be accepted as the 2nd one. 

### How to reproduce

need to add the `::date` in the min not in the case 
❌ **the problem**` , min(case when new_value_string = 'Lost' then event_history_at end)::date as lost_date`
✅  **no problem** `, min(case when new_value_string = 'Lost' then event_history_at::date end) as lost_date`

### Dialect

redshift

### Version

sqlfluff==2.1.0
sqlfluff-templater-dbt==2.1.0

### Configuration

```ini
[sqlfluff]

dialect = redshift

exclude_rules = L026, L031, L032
large_file_skip_byte_limit = 0

max_line_length = 120

[sqlfluff:layout:type:comma]
line_position = leading

[sqlfluff:rules:references.keywords]
ignore_words = name, year_month, state, type, stage, subscription, level, comment, event

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
# Unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
# Function names
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
# Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
# Data Types
extended_capitalisation_policy = lower

[sqlfluff:rules:aliasing.length]
min_alias_length = 3

[sqlfluff:rules:references.special_chars]
# Special characters in identifiers
quoted_identifiers_policy = none
ignore_words = url, referrer, name

[sqlfluff:rules:convention.casting_style]
# SQL type casting
preferred_type_casting_style = shorthand

[sqlfluff:indentation]
# See https://docs.sqlfluff.com/en/stable/layout.html#configuring-indent-locations
allow_implicit_indents = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by JohnLemmonMedely** [comment]
We just hit this bug and I wanted to add a bit more detail because rule AL03 [aliasing.expression] is also affected. 

It looks like sqlfluff is treating any column that's casted at the end as if it wasn't a column expression. Take this simple example.
```
select
    col_a,
    cast(col_a + col_b as float),
    col_a + col_b,
    (col_a + col_b)::float
from my_table
```

I get the original ST06 error from this ticket on the 4th column. I also get rule AL03 triggering for the 2nd and 3rd columns but not the 4th even though they use the same logic. "Column expression without alias. Use explicit `AS` clause. [aliasing.expression]"

One workaround is to use the `cast` function instead of the `::` syntax, at least with Snowflake.

**Comment by jonyscathe** [comment]
Any change that this issue will be worked on at some point.
I agree completely with the logic behind ST06, and hate having to put any column with a cast with the simple targets, rather than calculations/aggregates.

**Comment by peterbud** [comment]
The casting operator `::` is considered as a calculation. Let me see if we can exempt that.
